packages/block-editor/src/store/selectors.js

```diff
@@ -194,11 +194,9 @@ export const getBlocks = createSelector(
 			getBlock( state, clientId )
 		);
 	},
-	( state, rootClientId ) =>
-		map(
-			state.blocks.order[ rootClientId || '' ],
-			( id ) => state.blocks.cache[ id ]
```

Comment:
Is cache key not needed anymore or do we have two cache keys now? Why these two cache keys are needed, what's the purpose of each one of them?

Reply:
Yes, the main cache is for `getBlock` selector, but the new cache is for the `getBlocks` selector. 

Currently, `getBlocks` is computed from `getBlock` cache -- it's an array which contains the cache key for each child block. So it invalidates when any child changes.

The problem is that we compute that cache on the fly. So we have no ability to override it -- if a child block changes, getBlocks always invalidates.

This is problematic because there are times when we need the cache to stay the same for a specific action, even when it appears as if the child blocks are changing. For example, when you add a template part to the editor which is a duplicate of an existing template block, it will do `replaceInnerBlocks`. This first removes all the specified blocks and then adds them to the correct location. This is problematic because since both template parts pull from the same entity, all of the child blocks of each template part share the same clientIds.

So the "remove the blocks" part of `replaceInnerBlocks` actually removes them from the existing template part block, which causes that template part to update the root entity, which causes the duplicate template part to see that it needs to remove its blocks, so both template parts are effectively cleared of all content.

We can get around this by making sure that the `getBlocks` selector does not update during the `replaceInnerBlocks` action. And we can do that by keeping the cache key the same each run.

But to do that, we need to be able to control the cache key for `getBlocks` from the store, rather than just computing it on the selector level.

The `getBlock` selector still does the same thing as it did before.

Reply:
> For example, when you add a template part to the editor which is a duplicate of an existing template block, it will do replaceInnerBlocks. This first removes all the specified blocks and then adds them to the correct location.

I'm not sure I understand everything right but from my perspective, it's not the state this is incorrect or the selectors, but the action is not doing what it's supposed to do here.

> This first removes all the specified blocks and then adds them to the correct location

In  other words, why does this action remove the block and add them back if they are already present on state?

> We can get around this by making sure that the getBlocks selector does not update during the replaceInnerBlocks action. And we can do that by keeping the cache key the same each run.

When I read this sentence, it feels to me like we're going in a wrong direction. Maybe it's the only one because I don't  understand everything  yet but having a  specific behavior in a selector depending on what  the last action was  is problematic and can cause hidden bugs. In other words for me, `getBlocks` cache key being the concatenation  ofthe cache keys of all blocks is logic and playing with cache keys just to avoid having a selector change its value even if the state value actually changed is an issue. The cache key is just an optimization, if I remove the memoization, the selector should return the same value but as written in this PR, you're hacking the memoization to avoid running the selector again even if the state value really changed. (Correct me if I'm wrong)

Reply:
`replaceInnerBlocks` is working as intended unfortunately:

> Returns an action object used in signalling that the inner blocks with the specified client ID should be replaced.

Say I have `replaceInnerBlocks( 'template-part-id', blocks )`. This action does two things:
1. Remove all child blocks of `template-part-id`. 
2. Add all given `blocks` as children of `template-part-id`.

Say I two template parts pulling from the same data source, all of their chilldren bocks are going to share the same clientID:

```
templatePartA (using template part with ID 7):
  - block1
  - block2
  - block3
postContent
templatePartB (using template part with ID 7):
  - block1
  - block2
  - block3
```

because the template part with ID 7 is going to have the exact same blocks under it.

So when I insert `templatePartB`, it uses `replaceInnerBlocks` to load its blocks into the state. This means that all blocks are cleared from both `templatePartA` and `templatePartB`, which ends up triggering that update which deletes the blocks.

if I'm not mistaken, the single-block use case gets around this by keeping the cache key for the parent block being modified the same until the entire action has finished. I think that was my motivation for doing the same for `getBlocks`.

I think your analysis is spot on though about hacking memoization. I think we are unfortunately already doing this in a few places :/

I can think on some other ways to solve this. Possibly we can avoid `replaceInnerBlocks`, or we could try to keep track of multiple blocks using the same data source in `useBlockSync`.

Reply:
> So when I insert templatePartB, it uses replaceInnerBlocks to load its blocks into the state. This means that all blocks are cleared from both templatePartA and templatePartB, which ends up triggering that update which deletes the blocks.

This is where you lost me, why triggering `replaceInnerBlocks` for templatePartA clear the blocks for templatePartB?

Reply:
Reading this conversation and thinking about it, I came to realize that we make at least one wrong assumption in the current shape of the reducer state. We assume a block with a given ID has just a unique parent.

I think what we should think about is:

 - Do we want to fix that assumption in the state? what does it mean for the ancestor selectors we have right now: (basically what path to choose)
 - Alternatively and I think maybe this is safer,  we shouldn't reuse the same block client Id for controlled entities. i think there's a lot of things in the editor that assumes clientId is unique and breaking that assumption can be  dangerous in ways we don't see yet.

Reply:
So if template part A pulls from "entity1" and template part B also pulls from "entity1," all the blocks in entity1 have a specific client ID, right? That client ID doesn't change when inserting it into the block editor. So if my entity has blocks with IDs `[ 1, 2, 3 ]`, when I put those into the block editor, both template parts A and B have the blocks `[ 1, 2, 3 ]`. The client IDs are the same.

So when I do `replaceInnerBlocks` for templatePartA, I have to remove all the child blocks. The child blocks are `[ 1, 2, 3 ]`. So the remove blocks action is dispatched with those client IDs. So all the blocks with client Ids `[1, 2, 3]` are removed from the block editor store

Since templatePartB's children innerblocks are `[1, 2, 3]` because the entity is shared, it is referring to the exact same blocks in the store as templatePartA refers to. So since those blocks `[1, 2, 3]` are removed, they stop showing up under `templatePartB` as well

does that make sense? It's a pretty confusing situation. The block editor store is not really designed for there to one block with a certain clientId to render in two different places on the screen, but that is what's happening right now if you insert two template part blocks which both refer to the same template part entity.

Reply:
Actually, I wonder if we need some kind of "internalId" (the current clientId) and an "externalId" or  "clientId" specific to the entity.

Reply:
yeah that might be nice.

The benefit of sharing the clientID is that changes are updated instantaneously on all blocks with the same cilentId, which is kinda nice.

Reply:
> So when I do replaceInnerBlocks for templatePartA, I have to remove all the child blocks. The child blocks are [ 1, 2, 3 ]. So the remove blocks action is dispatched with those client IDs. So all the blocks with client Ids [1, 2, 3] are removed from the block editor store

The issue here is that we actually remove the blocks, we shouldn't if the same clientIds are being inserted or existed under another parent. (multi-parent again)

> The benefit of sharing the clientID is that changes are updated instantaneously on all blocks with the same cilentId, which is kinda nice.

The syncing mechanism could allow the changes to reflect with a very small delay even with different ids.

--- 
I think we should probably give some thoughts and maybe explore some fixes at the conceptual level (multi-parents or different clientIds) because I believe even if we fix the current issue with hacks (memoiziation or any other hack), if the conceptual ambiguity remains in the state, other bugs will surface. For instance I won't be surprised if dragging and dropping blocks become very weird if you have the same instance twice.

Reply:
> we shouldn't if the same clientIds are being inserted

That's true. We might be able to do some sort of comparison, but it's hard to think of a generic solution in this case that works for all instances of replaceInnerBlocks without some performance hit. I guess just comparing the top-level blocks could be good enough here 🤔 


The secondary ID thing is interesting since it would have to work both ways. So when we put blocks from an entity --> block editor, they get a proper unique ID, and going back the other way they need to go back to the clientID they had at the entity level.